include/ur_modern_driver/pipeline.h

```diff
@@ -141,27 +141,18 @@ class Pipeline
   {
     consumer_.setupConsumer();
     unique_ptr<T> product;
-    Time last_pkg = Clock::now();
-    Time last_warn = last_pkg;
     while (running_)
     {
-      // 16000us timeout was chosen because we should
-      // roughly recieve messages at 125hz which is every
-      // 8ms so double it for some error margin
-      if (!queue_.wait_dequeue_timed(product, std::chrono::milliseconds(16)))
+      // timeout was chosen because we should receive messages
+      // at roughly 125hz (every 8ms) and have to update
+      // the controllers (i.e. the consumer) with *at least* 125Hz
+      // So we update the consumer more frequently via onTimeout
+      if (!queue_.wait_dequeue_timed(product, std::chrono::milliseconds(8)))
       {
-        Time now = Clock::now();
-        auto pkg_diff = now - last_pkg;
-        auto warn_diff = now - last_warn;
-        if (pkg_diff > std::chrono::seconds(1) && warn_diff > std::chrono::seconds(1))
-        {
-          last_warn = now;
-          consumer_.onTimeout();
-        }
+        consumer_.onTimeout();
```

Comment:
The existing code here is a mess and it's no wonder you've removed it :-) 
Since `onTimeout` isn't used at all and it's sort of being abused as a heartbeat function in your code I'm wondering if it wouldn't make better sense to just always call `consume` after 8ms but in the case of failure, product will be set to `T::InvalidProduct` which then would have to be typedef'ed for `RTPacket`, `StatePacket` and `MessagePacket` together with extending the the consumers to accept the new packet types. 
At least that's something I might end up doing after merging if we can reach a conclusion for the Indigo problem.

Reply:
Both ways sound ok to me.
You are right that my proposal uses `onTimeout` as a heartbeat, but consuming a non-existing package as an invalid package is just as weird in my opinion.

It's simply a matter of where you want to place the guarantee on the lower-bound of the call-rate.

The important point here is that the current pipeline behavior breaks the minimum update rate and results in jerking trajectory execution